enhancements/authentication/service-ca-cert-generation-for-statefulset-pods.md

```diff
@@ -0,0 +1,344 @@
+The subject of `$podID.crt` is the pod’s unique DNS identity,
+`${statefulset.name}-${podID}.${service.name}.${service.namespace}.svc` .
+Each pod from a `StatefulSet` can then pick and use its own certificate.
+
+Creating a `StatefulSet` triggers both a `Secret` creation and creating the `StatefulSet` pods,
+which race against each other.
+So, each individual pod **must** fail if a certificate for its individual identity is not found.
+We also generate a few more certificate+key pairs than necessary,
+to allow scaling up without such pod failures.
+
+### User Stories
+
+#### Creating a TLS-protected StatefulSet
+
+- Author of the application, or of the container image wrapping the application,
+  ensures that the pod
+  - Determines its identity $podID (already the case)
+  - Loads `$podID.{crt.key}` on startup from a separate directory
+  - Exits with a failing status if those files are missing
+  - Uses the certificate+private key for accepting intra-StatefulSet connections
+  - Trusts the Service CA for making outgoing intra-StatefulSet connections,
+    using the signed `${statefulset.name}-${podID}.${service.name}.${service.namespace}.svc`
+    (or the same with `….cluster.local`) host names.
+- The StatefulSet for the application
+  is updated to contain:
+  - An annotation `service.beta.openshift.io/serving-cert-secret-name: $secretName`
+  - In the pod template, a volume referencing secret `$secretName`
+  - In the pod’s container template, a `volumeMount` mounting that volume
+    in the application-expected directory
+- The application administrator creates the StatefulSet object.
+- The StatefulSet controller starts creating pods for the StatefulSet.
+  They will initially fail because `$secretName` is missing or does not have the expected
+  `$podID.{crt,key}` files, causing the StatefulSet controller to wait and retry.
+- Concurrently, the Service CA controller observes the creation of the StatefulSet,
+  and creates a secret `$secretName` in the same namespace, with `{0,1,2,3,…}.{crt,key}` keys
+  matching the hostnames of the StatefulSet pods; the number of certificate/key pairs may be
+  larger than the requested number of replicas of the StatefulSet.
+- After the secret is created/updated, and the StatefulSet controller retries,
+  the pods start up succesfully.
+
+#### Scaling the StatefulSet up
+
+- An admin increases the StatefulSet’s requested replica count.
+- The StatefulSet controller starts creating the new pods.
+  If the certificate/key pairs already exist in the secret (e.g. if just addding one pod)
+  the pods start up succesfully;
+  if not, the pods fail, and the StatefulSet controller will wait and retry.
+- Concurrently, the Service CA controller observes the replica count increase;
+  if the existing secret does not have enough certificate/key pairs (including a margin),
+  it updates the secret so that it does.
+  (It’s unspecified whether pre-existing certificate/key pairs are reused or replaced.)
```

Comment:
This was intended a a starting point for end-user documentation: we make no commitment on that behavior (and because just updating a secret does not restart pods, it really should not matter to end users). I appreciate that this is not helpful for readers that want to understand what the feature actually does.

The _implementation_ currently always re-generates the whole set of certificates (because that’s simpler to implement, and so that they all expire at a similar time and we avoid secret churn on serial expirations).

Is there a better place for a draft of end-user documentation like this? Or maybe we should just document what the implementation currently does, and commit to that behavior, giving up a little flexibility for an easier-to-understand feature.